hibernate-core/src/test/java/org/hibernate/orm/test/hql/customFunctions/CountItemsGreaterValSqmFunction.java

```diff
@@ -0,0 +1,139 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.orm.test.hql.customFunctions;
+
+import org.hibernate.dialect.Dialect;
+import org.hibernate.dialect.function.CastFunction;
+import org.hibernate.metamodel.mapping.JdbcMapping;
+import org.hibernate.query.sqm.function.AbstractSqmSelfRenderingFunctionDescriptor;
+import org.hibernate.query.sqm.function.FunctionKind;
+import org.hibernate.query.sqm.produce.function.*;
+import org.hibernate.sql.ast.Clause;
+import org.hibernate.sql.ast.SqlAstTranslator;
+import org.hibernate.sql.ast.spi.SqlAppender;
+import org.hibernate.sql.ast.tree.SqlAstNode;
+import org.hibernate.sql.ast.tree.expression.CastTarget;
+import org.hibernate.sql.ast.tree.expression.Expression;
+import org.hibernate.sql.ast.tree.predicate.Predicate;
+import org.hibernate.type.BasicType;
+import org.hibernate.type.StandardBasicTypes;
+import org.hibernate.type.spi.TypeConfiguration;
+
+import java.math.BigDecimal;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.hibernate.query.sqm.produce.function.FunctionParameterType.NUMERIC;
+
+//tag::hql-user-defined-dialect-function-sqm-renderer[]
+public class CountItemsGreaterValSqmFunction extends AbstractSqmSelfRenderingFunctionDescriptor {
+	private final CastFunction castFunction;
+	private final BasicType<BigDecimal> bigDecimalType;
+
+	public CountItemsGreaterValSqmFunction(String name, Dialect dialect, TypeConfiguration typeConfiguration) {
+		super(
+			name,
+			FunctionKind.AGGREGATE,
+			/* Function consumes 2 numeric typed args:
+			- the aggregation argument
+			- the bottom edge for the count predicate*/
+			new ArgumentTypesValidator(StandardArgumentsValidators.exactly(2),
+					FunctionParameterType.NUMERIC,
+					FunctionParameterType.NUMERIC
+			),
+			// Function returns one value - the number of items
+			StandardFunctionReturnTypeResolvers.invariant(
+					typeConfiguration.getBasicTypeRegistry()
+							.resolve(StandardBasicTypes.BIG_INTEGER)
+			),
+			StandardFunctionArgumentTypeResolvers.invariant(
+					typeConfiguration, NUMERIC, NUMERIC
+			)
+		);
+		// Extracting cast function for setting input arguments to correct the type
+		castFunction = new CastFunction(
+				dialect,
+				dialect.getPreferredSqlTypeCodeForBoolean()
+		);
+		bigDecimalType = typeConfiguration.getBasicTypeRegistry()
+				.resolve(StandardBasicTypes.BIG_DECIMAL);
+	}
+
+	@Override
+	public void render(
+			SqlAppender sqlAppender,
+			List<? extends SqlAstNode> sqlAstArguments,
+			SqlAstTranslator<?> walker) {
+		render(sqlAppender, sqlAstArguments, null, walker);
+	}
+
+	//tag::hql-user-defined-dialect-function-sqm-renderer-definition[]
+	@Override
+	public void render(
+			SqlAppender sqlAppender,
+			List<? extends SqlAstNode> sqlAstArguments,
+			Predicate filter,
+			SqlAstTranslator<?> translator) {
```

Comment:
I'm really not sure how much of an open invitation we really want to be giving to users to start depending on some of these SPIs in their application programs.

Yes, it's true that `org.hibernate.sql.ast` and `org.hibernate.sql.ast.tree` are technically SPIs, but they're both marked `@Incubating` for the very good reason that:

1. they're an extremely broad SPI surface, and
2. not really stabilized in any meaningful sense.

So I would leave this undocumented for now.

Reply:
Let's close this and the Jira then.

I didn't find an issue about adding a proper API to define custom functions though. Should I create one, or is this not a feature we want to expose beyond an incubating SPI?

Reply:
> I didn't find an issue about adding a proper API to define custom functions though. 

I did have _something_ at some stage.